File: src/chainOfResponsibility/python/ChainLink.py
from abc import ABC, abstractmethod


class Link(ABC):
    @abstractmethod
    def check(self, username: str, password: str, token: str) -> bool:
        pass

    @abstractmethod
    def next(self, username: str, password: str, token: str) -> bool:
        pass


class ChainLink(Link, ABC):
    nextLink: Link = None

    @staticmethod
    def initialize(head: Link, *chain: Link):
        current: Link = head
        for n in chain:
            current.nextLink = n
            current = n

        return head

    def next(self, username: str, password: str, token: str) -> bool:
        if self.nextLink is not None:
            return self.nextLink.check(username, password, token)

        return True

File: src/chainOfResponsibility/java/com/my/ChainLink.java
package src.chainOfResponsibility.java.com.my;

public abstract class ChainLink {
    private ChainLink nextLink;

    public static ChainLink initialize(ChainLink head, ChainLink... chain) {
        ChainLink current = head;
        for (ChainLink next : chain) {
            current.nextLink = next;
            current = next;
        }

        return head;
    }

    public abstract boolean check(String username, String password, String token);

    protected boolean next(String username, String password, String token) {
        if (nextLink != null) {
            return nextLink.check(username, password, token);
        }

        return true;
    }
}
